File: java/src/ch1/ex115.java
package ch1;

public class ex115 {
    public static void main(String[] args) {
        System.out.println(sine(0));
        System.out.println(sine(1));
        System.out.println(sine(2));
        System.out.println(sine(3));
        System.out.println(sine(Math.PI / 6));
        System.out.println("done");
    }

    static double p(double x) {
        return 3 * x - 4 * cube(x);
    }

    static double sine(double angle) {
        if (Math.abs(angle) < 0.1) return angle;
        return p(sine(angle / 3.0));
    }

    static double cube(double x) {
        return x * x * x;
    }
}

File: haskell/ch1/Ex115.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load "ExXXX.hs"
-- Run with count_change 100 (should equal 292)

module Ex115 where

p x = 3 * x - 4 * cube x

sine angle | abs angle < 0.1 = angle
sine angle = p (sine (angle / 3.0))

cube x = x * x * x
